passes/cmds/exec.cc

```diff
@@ -0,0 +1,190 @@
+/*
+ *  yosys -- Yosys Open SYnthesis Suite
+ *
+ *  Copyright (C) 2012  Clifford Wolf <clifford@clifford.at>
+ *
+ *  Permission to use, copy, modify, and/or distribute this software for any
+ *  purpose with or without fee is hereby granted, provided that the above
+ *  copyright notice and this permission notice appear in all copies.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ *  WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ *  MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ *  ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ *  WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ *  ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ *  OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *
+ */
+
+#include "kernel/register.h"
+#include "kernel/log.h"
+#include <cstdio>
+#include <sys/wait.h>
+
+USING_YOSYS_NAMESPACE
+PRIVATE_NAMESPACE_BEGIN
+
+struct ExecPass : public Pass {
+	ExecPass() : Pass("exec", "execute commands in the operating system shell") { }
+	void help() YS_OVERRIDE
+	{
+		//   |---v---|---v---|---v---|---v---|---v---|---v---|---v---|---v---|---v---|---v---|
+		log("\n");
+		log("    exec [options] -- [command]\n");
+		log("\n");
+		log("Execute a command in the operating system shell.  All supplied arguments are\n");
+		log("concatenated and passed as a command to popen(3).  Whitespace is not guaranteed\n");
+		log("to be preserved, even if quoted.  stdin and stderr are not connected, while stdout is\n");
+		log("logged unless the \"-q\" option is specified.\n");
+		log("\n");
+		log("\n");
+		log("    -q\n");
+		log("        Suppress stdout and stderr from subprocess\n");
+		log("\n");
+		log("    -expect-return <int>\n");
+		log("        Generate an error if popen() does not return specified value.\n");
+		log("        May only be specified once; the final specified value is controlling\n");
+		log("        if specified multiple times.\n");
+		log("\n");
+		log("    -expect-stdout <regex>\n");
+		log("        Generate an error if the specified regex does not match any line\n");
+		log("        in subprocess's stdout.  May be specified multiple times.\n");
+		log("\n");
+		log("    -not-expect-stdout <regex>\n");
+		log("        Generate an error if the specified regex matches any line\n");
+		log("        in subprocess's stdout.  May be specified multiple times.\n");
+		log("\n");
+		log("\n");
+		log("    Example: exec -q -expect-return 0 -- echo \"bananapie\" | grep \"nana\"\n");
+		log("\n");
+		log("\n");
+	}
+	void execute(std::vector<std::string> args, RTLIL::Design *design) YS_OVERRIDE
+	{
+		std::string cmd = "";
+		char buf[4096] = {};
```

Comment:
Guess making buffer this big is not needed. Usually output of executable will be quite limited, and later you anyway loop in case it is larger then buffer. Understand this is making it more performant but not sure if needed.

Reply:
OK, sure, I'll change to 1KiB.  I don't think it makes much difference either way.